### PR TITLE
Allow agnostic NMS in validation for OBB, Pose, Segment and NAS

### LINKED ISSUE #16159: agnostic_nms=True is not working in yolov8-obb
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I trained normal and defect shapes using the Yolov8-OBB Model. However, there was a phenomenon that 'normal' and 'defect' classes were all detected in one object as in the example below, perhaps because agnostic_nms is not working when it is OBB. What should I do?

image

code
```python
import os
from ultralytics import YOLO, settings

# Setting environment variables for MLflow
os.environ["MLFLOW_EXPERIMENT_NAME"] = ""
os.environ["MLFLOW_RUN"] = ""

# Use MLflow
settings.update({"mlflow": True})
# settings.update({"cos_lr": True})

# Load a model
model = YOLO("yolov8m-obb.pt")

# Train the model
results = model.train(data="cfg/molding.yaml",
                      epochs=100,
                      imgsz=960,
                      rect=True,
                      single_cls=False,
                      close_mosaic=0,
                      augment=False,
                      hsv_h=0,
                      hsv_s=0,
                      hsv_v=0,
                      degrees=0,
                      translate=0,
                      scale=0,
                      shear=0,
                      perspective=0,
                      flipud=True,
                      fliplr=True,
                      mosaic=False,
                      mixup=False,
                      cls=1,
                      # box=4,
                      agnostic_nms=True,
                      freeze=12,
                      # lr0=0.00025,
                      # optimizer='Adam',
                      device=0)
```

### Additional

_No response_

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes https://github.com/ultralytics/ultralytics/issues/16159

I'm not sure why it is allowed for `DetectionValidator`
https://github.com/ultralytics/ultralytics/blob/5022c4bb508cce701f692e8d9c8eb8a9039799cd/ultralytics/models/yolo/detect/val.py#L100

but not for the rest:
https://github.com/ultralytics/ultralytics/blob/5022c4bb508cce701f692e8d9c8eb8a9039799cd/ultralytics/models/nas/val.py#L47

https://github.com/ultralytics/ultralytics/blob/5022c4bb508cce701f692e8d9c8eb8a9039799cd/ultralytics/models/yolo/segment/val.py#L79

https://github.com/ultralytics/ultralytics/blob/5022c4bb508cce701f692e8d9c8eb8a9039799cd/ultralytics/models/yolo/pose/val.py#L72

https://github.com/ultralytics/ultralytics/blob/5022c4bb508cce701f692e8d9c8eb8a9039799cd/ultralytics/models/yolo/obb/val.py#L48

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Added flexibility in detection by updating agnostic behavior in model postprocessing.

### 📊 Key Changes
- Modified `agnostic` parameter in several model postprocessing functions to include `agnostic_nms` option.

### 🎯 Purpose & Impact
- **Enhanced Flexibility**: Allowing `agnostic_nms` helps handle cases with more robust detection across different classes.
- **Improved Accuracy**: Users may experience better detection results in scenarios where class-specific distinctions aren't crucial.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements to non-maximum suppression (NMS) options across validation scripts for different models.

### 📊 Key Changes
- Modified NMS logic in multiple model validation scripts (`nas`, `obb`, `pose`, `segment`) to use both `single_cls` and `agnostic_nms` parameters.

### 🎯 Purpose & Impact
- **Improved Flexibility**: Provides enhanced control over classification settings during evaluation by allowing combination of `single_cls` and `agnostic_nms`.
- **Enhanced Model Testing**: Ensures more robust testing capabilities across different model types, potentially leading to improved accuracy and performance insights.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a PR to the Ultralytics 🚀 repository! Here's some guidance to help ensure a smooth integration:

- ✅ Ensure your PR is **up-to-date** with the `main` branch of `ultralytics/ultralytics`. You can update your branch by clicking 'Update branch' or by running `git pull` and `git merge main` locally.
- ✅ Make sure all Continuous Integration (CI) **checks are passing**.
- ✅ Keep your changes to the bare **minimum** needed for your update. This helps maintain clarity and focuses on the specific improvements you're making.
- ✅ Update the Ultralytics [Documentation](https://docs.ultralytics.com) if your changes include any new or modified functionality.

For more information on how to contribute effectively, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Feel free to ask if you have any questions!

🛠️ Your changes in this PR add useful flexibility by allowing agnostic NMS across various validators. This enhancement should improve detection capabilities in scenarios where class specificity is less critical. Great work!

A human reviewer will respond soon with additional feedback. Thank you for your contribution! 🎉

**Comment by glenn-jocher** [comment]
@Y-T-G oh wow, great fixes!

**Comment by glenn-jocher** [comment]
@Y-T-G PR merged!
